File: Python/UKF/UTs/UT.py
import numpy as np
import math as mth
import sys 
sys.path.append('..')
from Auxiliar_Functions.auxiliar_functions import *
from Sigma_Representations.sigma_representations import *

#Unscented transformation implementation: ut
#Author: Victor Araujo Vieira.
#e-mail: icevct@gmail.com
#University of Brasilia - Brazil.

#ut function definition
#As there's so many variables, the arrays dimensions were not specified
#Inputs: x_previous (Array)
#		 pxx_previous (Array)
#        f (A function. The state or measurement state function)
#        param_f (Array or an empty list [])
#        weight (scalar)
#		 nargout (scalar): the variable that controls how many 
#							values the function will return
#        *args (scalar): It can be 1 or 2 argument, which will be the option to which sigma represetantion will be used or 
#						 dim_y a scalar, that indicates the amount of measurements 
#Outputs: x_predicted (Array)
#		  p_predicted (Array)
#		  p_cross_covariance (Array)
def ut(x_previous, pxx_previous, f, param_f, weight, nargout, *args):
	
	if(len(args) == 1):
		n = x_previous.shape[0]; # amount of states
	else:
		n = args[1] #amount of measurements

	#Chooses which sigma representation will be used, based on the option passed as argument
	if(args[0] == 0):
	#even_homi_sy_sig_rep
	    sigma_points_before, weights = even_homi_sy_sig_rep(x_previous, pxx_previous)
	elif(args[0] == 1):
	#sig_rep_julier_1995
	    sigma_points_before, weights = sig_rep_julier_1995(x_previous, pxx_previous, weight)
	elif(args[0] == 2):
	#ho_mi_sy_sig_rep
	    sigma_points_before, weights = ho_mi_sy_sig_rep(x_previous, pxx_previous, weight)
	elif(args[0] == 3):
    #rho_mi_sig_rep
	    sigma_points_before, weights = rho_mi_sig_rep(x_previous, pxx_previous, weight)
	#elif(args[0] == 4):
	else:
	    raise ValueError("There's not that option for a Sigma Representation!")


	N = sigma_points_before.shape[1] #amount of weights and sigma points columns 
	sigma_points_now = np.zeros((n, N)) 
	#sigma points of current sample, evaluating the sigma points to the state or measurement function

	for i in range(0, N):
		if(param_f == []):
			sigma_points_now[:, [i]] = feval(f, sigma_points_before[:, [i]])
		else:
			sigma_points_now[:, [i]] = feval(f, sigma_points_before[:, [i]], param_f)

	#function that calculates the x_predicted and p_predicted
	x_predicted, p_predicted = state_or_measure_predicted_cov_matrix_predicted(sigma_points_now, weights)

	#function that generates the cross covariance matrix
	p_cross_covariance = p_crosscovariance_with_weights(sigma_points_before, sigma_points_now, weights)

	#if nargout = 2, that means that there's only 2 return values
	#if nargout = 3, that means that there's 3 return values
	#the AdUKF function (main) will deal with it
	if nargout == 2:
		return x_predicted, p_predicted
	elif nargout == 3:
		return x_predicted, p_predicted, p_cross_covariance

File: Matlab/UKF/UTs_SigmaRep/UTs/UT.m
function [Xpredicted, Ppredicted, Pcrosscovariance] = UT(X_Previous, Pxx_Previous, f, param_f, Weight, Sigma_Rep, Measur_Dim)
%% Description
%Implementation of the Unscented Transformation function

%Author: Victor Araujo Vieira.
%e-mail: icevct@gmail.com
%University of Brasilia - Brazil.

%Inputs
%X_Previous: Initial State matrix
%Pxx_Previous: Covariance matrix
%f: The state function
%param_f: The state's function parameters
%Weight: The weight that was handled to the main UKF function
%Sigma_Rep: The option parameter, which chooses the sigma representation
%           that will be used
%Measur_Dim: The amount of rows in the measurements function

%Outputs
%Xpredicted: The predicted state
%Ppredicted: The predicted covariance matrix
%Pcrosscovariance: The cross covariance matrix

%% Implementation

%n: amount of states or measurements
if(nargin < 7)
    n = size(X_Previous, 1);
else
    n = Measur_Dim;
end

%Functions that computes the sigma points
%Sigma_Rep its a option variable, and it will choose the sigma
%representation that will be used
switch Sigma_Rep
    case 0
        [SigmaPointsBefore, Weights] = EvenHomiSySigRep(X_Previous, Pxx_Previous);
    case 1
        [SigmaPointsBefore, Weights] = SigRepJulier1995(X_Previous, Pxx_Previous, Weight);
    case 2
        [SigmaPointsBefore, Weights] = HoMiSySigRep(X_Previous, Pxx_Previous, Weight);
    case 3
        [SigmaPointsBefore, Weights] = RhoMiSigRep(X_Previous, Pxx_Previous, Weight);
    otherwise
         error('There are not that option for Sigma Representation!');
end

N = size(SigmaPointsBefore,2); %n amount of sigma points rows and weights

%Generates the current sample sigma pointsm evaluating the previous sigma
%points computed by the sigma representation function
SigmaPointsNow = zeros(n, N);
for i=1:N
    if(isempty(param_f))
        SigmaPointsNow(:, i) = feval(f , SigmaPointsBefore(:, i));
    else
        SigmaPointsNow(:, i) = feval(f , SigmaPointsBefore(:, i), param_f);
    end
end

%Function that computes the XPredicted and the Ppredicted
[Xpredicted, Ppredicted] = State_or_MeasurePredicted_CovMatrixPredicted(SigmaPointsNow, Weights);

%Function that computes the cross covariance matrix Pxy
Pcrosscovariance = Pcrosscovariance_withWeights(SigmaPointsBefore, SigmaPointsNow, Weights);

end